exporter/internal/queue/default_batcher.go

```diff
@@ -42,35 +43,73 @@ func (qb *DefaultBatcher) startReadingFlushingGoroutine() {
 			}
 
 			qb.currentBatchMu.Lock()
-			if qb.currentBatch == nil || qb.currentBatch.req == nil {
-				qb.resetTimer()
-				qb.currentBatch = &batch{
-					req:     req,
-					ctx:     ctx,
-					idxList: []uint64{idx}}
-			} else {
-				mergedReq, mergeErr := qb.currentBatch.req.Merge(qb.currentBatch.ctx, req)
-				if mergeErr != nil {
-					qb.queue.OnProcessingFinished(idx, mergeErr)
+
+			if qb.batchCfg.MaxSizeItems > 0 {
+				var reqList []internal.Request
+				var mergeSplitErr error
+				if qb.currentBatch == nil || qb.currentBatch.req == nil {
+					qb.resetTimer()
+					reqList, mergeSplitErr = req.MergeSplit(ctx, qb.batchCfg.MaxSizeConfig, nil)
+				} else {
+					reqList, mergeSplitErr = qb.currentBatch.req.MergeSplit(ctx, qb.batchCfg.MaxSizeConfig, req)
+				}
+
+				if mergeSplitErr != nil || reqList == nil {
+					qb.queue.OnProcessingFinished(idx, mergeSplitErr)
 					qb.currentBatchMu.Unlock()
 					continue
 				}
-				qb.currentBatch = &batch{
-					req:     mergedReq,
-					ctx:     qb.currentBatch.ctx,
-					idxList: append(qb.currentBatch.idxList, idx)}
-			}
 
-			if qb.currentBatch.req.ItemsCount() > qb.batchCfg.MinSizeItems {
-				batchToFlush := *qb.currentBatch
-				qb.currentBatch = nil
-				qb.currentBatchMu.Unlock()
-
-				// flushAsync() blocks until successfully started a goroutine for flushing.
-				qb.flushAsync(batchToFlush)
-				qb.resetTimer()
+				if reqList[0].ItemsCount() >= qb.batchCfg.MinSizeItems {
+					// If the incoming request is large enough that we decide to flush immediately, we flush all parts.
+					qb.currentBatch = nil
+					qb.currentBatchMu.Unlock()
+					for i := 0; i < len(reqList); i++ {
+						qb.flushAsync(batch{
+							req:     reqList[i],
+							ctx:     ctx,
+							idxList: []uint64{idx}})
+						// TODO: handle partial failure
+					}
+					qb.resetTimer()
+				} else {
+					qb.currentBatch = &batch{
+						req:     reqList[0],
```

Comment:
It looks like we should. Otherwise, we just ignore other items potentially. This situation shouldn't be allowed typically. However, I think we should either make a best effort here or throw a warning.